### PR TITLE
Add more optional information for RegFieldDesc to align with IP-XACT

### PR DESCRIPTION
**Pull request by mwachs5**
IP-XACT has the concept of "AddressBlock" and "RegisterFile". Allow users to annotate their RegFieldDesc 's with AddressBlock information and start exporting it to the Object Model. 

In this PR I believe I have done the necessary work for the AddressBlock new feature up to making the Object Model actually emit it. Note that RegisterFile's recursive nature made it too complicated and it did not end up in this PR.

**Related issue**: 

See discussion on https://github.com/sifive/duh-scala/issues/79#issuecomment-617485931

**Type of change**: feature request

**Impact**:API addition (no impact on existing code)

**Development Phase**:  implementation

**Release Notes**

Add the ability to specify AddressBlocks for RegFieldDesc and have them show up in the Object Model.

### CODE REVIEW

File: src/main/scala/diplomaticobjectmodel/model/OMAddressing.scala [hunk 1]
```diff
@@ -63,6 +65,22 @@ case class OMRegFieldGroup (
   _types: Seq[String] = Seq("OMRegFieldGroup", "OMCompoundType")
 ) extends OMCompoundType
 
+case class OMAddressBlock (
+  name: String,
+  addressOffset:  BigInt,
+  range: BigInt,
+  size: Int,
```
Comment (mwachs5) [inline src/main/scala/diplomaticobjectmodel/model/OMAddressing.scala:72]:
```suggestion
  width: Int,
```

File: src/main/scala/diplomaticobjectmodel/model/OMAddressing.scala [hunk 2]
```diff
@@ -63,6 +65,22 @@ case class OMRegFieldGroup (
   _types: Seq[String] = Seq("OMRegFieldGroup", "OMCompoundType")
 ) extends OMCompoundType
 
+case class OMAddressBlock (
+  name: String,
+  addressOffset:  BigInt,
```
Comment (richardxia) [inline src/main/scala/diplomaticobjectmodel/model/OMAddressing.scala:70]: This should be `baseAddress` to match the [duh](https://github.com/sifive/duh-schema/blob/4cf25992b38bdece9ecde92fe3e864713bbe3eef/lib/memoryMap.js#L24) and [IP-XACT](http://www.accellera.org/XMLSchema/IPXACT/1685-2014/memoryMap.xsd) terminology.

**Comment by richardxia** [comment]
> Should we just focus on the AddressBlock in this PR?

I'm fine with just focusing on the AddressBlock for now.

**Comment by drom** [comment]
LGTM

File: src/main/scala/diplomaticobjectmodel/model/OMAddressing.scala [hunk 3]
```diff
@@ -48,6 +48,8 @@ case class OMRegFieldDesc(
   volatile: Boolean,
   resetValue: Option[BigInt],
   enumerations: Seq[OMRegFieldEnumeration] = Seq(),
+  addressBlock: Option[String] = None,
+  // TODO: register files?
```
Comment (mwachs5) [inline src/main/scala/diplomaticobjectmodel/model/OMAddressing.scala:52]:
```suggestion
  // TODO: register files
```

File: src/main/scala/regmapper/RegFieldDesc.scala [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  // Enumerated values that this register field can take
+  enumerations: Map[BigInt, (String, String)] = Map(),
+  // The IP-XACT concept of an addressBlock which this
+  // register field can considered to be a part of, if exporting
+  // IP-XACT or similar outputs. 
+  addressBlock: Option[AddressBlockInfo] = None,
+  // The IP-XACT concept of a register file (recursively)
+  // which this register field can be considered to be a part of,
+  // if exporting IP-XACT or similar outputs.
+  regFiles: Seq[RegFileInfo] = Nil
```
Comment (mwachs5) [inline src/main/scala/regmapper/RegFieldDesc.scala:69]:
```suggestion
  addressBlock: Option[AddressBlockInfo] = None
  // TODO: registerFiles
```

File: src/main/scala/regmapper/RegFieldDesc.scala [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+// this has no effect.
+object RegFieldRegFiles {
+  def apply (regFileInfos: Seq[RegFileInfo], regs: Seq[RegField]): Seq[RegField] = {
+    regs.map {r =>
+      r.desc.map { d =>
+        r.copy(desc = Some(d.copy(regFiles = regFileInfos)))
+      }.getOrElse(r)
+    }
+  }
+}
```
Comment (mwachs5) [inline src/main/scala/regmapper/RegFieldDesc.scala:151]:
```suggestion
// TODO: RegisterFiles
```

**Comment by mwachs5** [comment]
@richardxia can you please approve? @drom approval is unfortunately not enough.

**Pull request merged**